fundamentals/android/auth-sign-in-with-google/Notes/app/src/main/java/com/notes/app/screens/authentication/CredentialsExt.kt

```diff
@@ -0,0 +1,115 @@
+package com.notes.app.screens.authentication
+
+import android.content.Context
+import android.util.Log
+import androidx.compose.foundation.layout.fillMaxWidth
+import androidx.compose.foundation.layout.padding
+import androidx.compose.material3.Button
+import androidx.compose.material3.ButtonDefaults
+import androidx.compose.material3.Icon
+import androidx.compose.material3.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.rememberCoroutineScope
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.platform.LocalContext
+import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.unit.sp
+import androidx.credentials.Credential
+import androidx.credentials.CredentialManager
+import androidx.credentials.GetCredentialRequest
+import androidx.credentials.exceptions.GetCredentialException
+import androidx.credentials.exceptions.NoCredentialException
+import com.google.android.libraries.identity.googleid.GetGoogleIdOption
+import com.google.android.libraries.identity.googleid.GetSignInWithGoogleOption
+import com.notes.app.ERROR_TAG
+import com.notes.app.R
+import com.notes.app.SnackbarManager
+import com.notes.app.ui.theme.Purple40
+import kotlinx.coroutines.launch
+
+@Composable
+fun AuthenticationButton(buttonText: Int, onRequestResult: (Credential) -> Unit) {
+  val context = LocalContext.current
+  val coroutineScope = rememberCoroutineScope()
+
+  Button(
+    onClick = { coroutineScope.launch { launchCredManButtonUI(context, onRequestResult) } },
+    colors = ButtonDefaults.buttonColors(containerColor = Purple40),
+    modifier = Modifier
+      .fillMaxWidth()
+      .padding(16.dp, 0.dp)
+  ) {
+    Icon(
+      painter = painterResource(id = R.drawable.google_g),
+      modifier = Modifier.padding(horizontal = 16.dp),
+      contentDescription = "Google logo"
+    )
+
+    Text(
+      text = stringResource(buttonText),
+      fontSize = 16.sp,
+      modifier = Modifier.padding(0.dp, 6.dp)
+    )
+  }
+}
+
+private suspend fun launchCredManButtonUI(
+  context: Context,
+  onRequestResult: (Credential) -> Unit
+) {
```

Comment:
Instead of having a callback here, what do you think of returning the result?
It seems like most code here can be run synchronously, I don't see anything requiring a callback.

Reply:
Oh, unless this code displays some type of UI and awaits user input?

Reply:
Yes, it waits for the user input. Once the user clicks on the account they'd like to use, it returns the chosen credential, which we then pass to the ViewModel through this callback .